src/python_testing/TC_WEBRTCP_2_16.py

```diff
@@ -0,0 +1,180 @@
+        return "[TC-WEBRTCP-2.16] Validate ProvideOffer resource exhaustion - PROVISIONAL"
+
+    def steps_TC_WEBRTCP_2_16(self) -> list[TestStep]:
+        steps = [
+            TestStep(1, "TH allocates both Audio and Video streams via AudioStreamAllocate and VideoStreamAllocate commands to CameraAVStreamManagement",
+                     "DUT responds with success and provides stream IDs"),
+            TestStep(2, "TH sends multiple ProvideOffer commands to exhaust DUT's session capacity (DUT-specific limit)",
+                     "DUT responds with ProvideOfferResponse for each until capacity is reached"),
+            TestStep(3, "TH sends an additional ProvideOffer command beyond DUT's capacity",
+                     "DUT responds with RESOURCE_EXHAUSTED status code"),
+        ]
+        return steps
+
+    def pics_TC_WEBRTCP_2_16(self) -> list[str]:
+        pics = [
+            "WEBRTCP.S",
+            "AVSM.S",
+        ]
+        return pics
+
+    @async_test_body
+    async def test_TC_WEBRTCP_2_16(self):
+        """
+        Executes the test steps for validating ProvideOffer resource exhaustion.
+        """
+
+        endpoint = self.get_endpoint(default=1)
+
+        self.step(1)
+        # Allocate Audio and Video streams
+        audio_stream_id = await self.allocate_one_audio_stream()
+        video_stream_id = await self.allocate_one_video_stream()
+
+        # Validate that the streams were allocated successfully
+        await self.validate_allocated_audio_stream(audio_stream_id)
+        await self.validate_allocated_video_stream(video_stream_id)
+
+        self.step(2)
+        # Send multiple ProvideOffer commands to exhaust the DUT's capacity
+        logger.info("Starting to send ProvideOffer commands to exhaust DUT capacity")
+
+        # Get the maximum concurrent WebRTC sessions from user or use default for CI
+        prompt_msg = (
+            "\nPlease enter the maximum number of concurrent WebRTC sessions supported by the DUT:\n"
+            "This value is DUT-specific and should be obtained from the DUT documentation or specifications.\n"
+            "Enter the number (e.g., 5): "
+        )
+
+        if self.is_pics_sdk_ci_only:
+            # Use default value for CI testing
+            max_attempts = 5
+            logger.info(f"Using default max_attempts={max_attempts} for CI testing")
+        else:
+            # Prompt user for DUT-specific value
+            user_input = self.wait_for_user_input(prompt_msg)
+            try:
+                max_attempts = int(user_input.strip())
+                asserts.assert_true(max_attempts > 0, "Maximum concurrent sessions must be greater than 0")
+                logger.info(f"Using user-specified max_attempts={max_attempts}")
+            except ValueError:
+                asserts.fail(f"Invalid input '{user_input}'. Please enter a valid number.")
+
+        provide_offer_cmd = Clusters.WebRTCTransportProvider.Commands.ProvideOffer(
```

Comment:
Can be done in follow-up, but does it make sense to move some of this command-sending, which is used throughout the test scripts, to a helper? See e.g. https://github.com/project-chip/connectedhomeip/blob/8abc4a68b62e9aa1d932f1dc8da78480d87f5f52/src/python_testing/TC_TLS_Utils.py#L46